src/main/java/org/mockito/internal/stubbing/defaultanswers/ReturnsDeepStubs.java

```diff
@@ -48,6 +53,10 @@ public class ReturnsDeepStubs implements Answer<Object>, Serializable {
 
     @Override
     public Object answer(InvocationOnMock invocation) throws Throwable {
```

Comment:
In this answer, on line 68 we already call `delegate().answer()` where the delegate is `ReturnsEmptyValues`. This class in turn already returns an empty optional: https://github.com/mockito/mockito/blob/076e8ac837c17e81347854feaad69687cd7d2b83/src/main/java/org/mockito/internal/stubbing/defaultanswers/ReturnsEmptyValues.java#L142-L149

Is there a different reason why we are not hitting that codepath?

Reply:
1. Because `Optional` returns to be mockable in line 66 `!typeMockabilityOf(rawType, mockSettings.getMockMaker()).mockable()`.
2. The `delegate().answer()` would return a real `Optional`, which was mentioned in the ticket not to be the right thing to return. As the `RETURNS_DEEP_STUBS` also do not return `delegate().answer()` for e.g. `Lists`.

Reply:
In that case I would say the delegation is not working as expected and we should fix that instead. I would like us not to reimplement the logic of returning empty values in another place, which is why we have the delegate. If that is not working properly, I consider that the bug. Can you look into fixing that while we make sure we don't break other use cases?

Reply:
@TimvdLippe Yes I can do that.
But just to be sure, that will also return a real Optional in the `RETURNS_DEEP_STUBS` case.
So the observable behavior for the client will be different for e.g. Lists and Optionals, when using  `RETURNS_DEEP_STUBS`.

Reply:
@TimvdLippe done

Reply:
Are we sure this is what we want?

This would make the default behaviour consistent, i.e. returning an empty optional, but would change the way users would need to deep mock methods returning optionals (and collections?).  Instead of mocking the calls to the optional's methods, users would need to mock the method returning the optional.

For example:

```java
when(a.getOptional().isPresent()).thenReturn(true);
when(a.getOptional().get()).thenReturn("thing");
```

Would need to become:

```java
when(a.getOptional()).thenReturn(Optional.of("thing"));
```

Or, if they really did still want an optional mock for some reason:

```java
Optional<String> mockOptional = mock(Optional.class);
when(mockOptional.isPresent()).thenReturn(true);
when(mockOptional.get()).thenReturn("thing");
when(a.getOptional()).thenReturn(mockOptional);
```

Personally, I think returning a real empty optional is better as the optional is an immutable object, so returning an immutable instance means the immutability is maintained in testing by default, highlighting unreachable branches etc.

It is a change that would need documenting both in the standard docs and in the release notes to help users upgrade (it's a breaking change).

Reply:
@TimvdLippe The proposed change to use `ReturnsEmptyValues` breaks multiple existing (>=12) tests.
Shall I go forward and change these tests to comply to the new behavior?
This seams like an invasive change.

Example:
`org.mockitousage.basicapi.MocksSerializationTest#BUG_ISSUE_399_try_some_mocks_with_current_answers`
`org.mockito.internal.stubbing.defaultanswers.ReturnsGenericDeepStubsTest#will_return_default_value_on_non_mockable_nested_generic`

and more...

Reply:
Hm yes that is too many breakages unfortunately. Still, I would prefer us to not implement this same list twice. Can we come up with a way that uses both?

Reply:
I could delegate the check into `ReturnsEmptyValue` where I extract the Optional checks into an own method, then we have the different logic, but near each other to see the differences.

Reply:
@TimvdLippe Done.